src/audio/pcm_converter/pcm_converter_generic.c

```diff
@@ -673,6 +676,71 @@ static int pcm_convert_s24_c32_to_s16_c32(const struct audio_stream *source,
 	return samples;
 }
 #endif
+
+#if CONFIG_PCM_CONVERTER_FORMAT_S24_C24_AND_S24_C32
+static int pcm_convert_s24_c24_to_s24_c32(const struct audio_stream *source,
+					  uint32_t ioffset, struct audio_stream *sink,
+					  uint32_t ooffset, uint32_t samples)
+{
+	uint8_t *src = source->r_ptr;
+	uint32_t *dst = sink->w_ptr;
+	int processed;
+	int nmax, i, n;
+
+	src += ioffset * 3;
+	dst += ooffset;
+	for (processed = 0; processed < samples; processed += n) {
+		src = audio_stream_wrap(source, src);
+		dst = audio_stream_wrap(sink, dst);
+		n = samples - processed;
+		nmax = audio_stream_bytes_without_wrap(source, src) / 3;
+		n = MIN(n, nmax);
+		nmax = audio_stream_bytes_without_wrap(sink, dst) >> BYTES_TO_S32_SAMPLES;
+		n = MIN(n, nmax);
+		for (i = 0; i < n; i += 1) {
+			*dst = (*(src + 2) << 16) | (*(src + 1) << 8) | *(src + 0);
```

Comment:
I think it would be good place this into a int32_t temporary variable and shift the bytes to bit32 (MSB aligned). Then right shift by 8 to create the sign extension. I don't think all components handle S24 correctly if the sign extension is missing.

Reply:
It is byte reorder. The original S24_3LE has the sign extension in *(src+2) bit 23, now it will be kept in S24_4LE bit 23

Reply:
@singalsu  For S24_4LE, the highest 8bit will be zero ?

Reply:
Actually I first used int32_t but I got some noise and spent a few days to find it should be uint32_t

Reply:
You should do `int32_t tmp = (*(src + 2) << 24) | (*(src + 1) << 16) | (*(src + 0) << 8); *dst = tmp >> 8;`

Reply:
The s24_4le format is not well defined. I can be with out without sign extension (where bits 31..24 are copy of bit 23). By not  having sign extension there's a risk that some component uses it for arithmetic in int32_t type without making sign extension that will cause strange audio waveforms. I think it's better to be safe and do the sign extension since it comes here with very low overhead.

Reply:
More ... also having the sign extension helps with C code debugger, you can see e.g. than PCM sample value is -1 and not 16777215.

Reply:
sure, I will try and validate it on windows

Reply:
@RanderWang How did it go, did it work?

Reply:
yes, your idea works. But I have another problem that the conversion algorithm is different between gateway module and normal module. I am working on it, .e.g. hda dma or alh dma require different layout for hw feature.